smart_open/ftp.py

```diff
@@ -0,0 +1,91 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+
+"""Implements I/O streams over FTP.
+"""
+
+import getpass
+import logging
+import urllib.parse
+import smart_open.utils
+
+logger = logging.getLogger(__name__)
+
+SCHEME = "ftp"
+
+"""Supported URL schemes."""
+
+DEFAULT_PORT = 21
+
+URI_EXAMPLES = (
+    'ftp://username@host/path/file',
+    'ftp://username:password@host/path/file',
+    'ftp://username:password@host:port/path/file'
+)
+
+
+def _unquote(text):
+    return text and urllib.parse.unquote(text)
+
+
+def parse_uri(uri_as_string):
+    split_uri = urllib.parse.urlsplit(uri_as_string)
+    assert split_uri.scheme in SCHEME
+    return dict(
+        scheme=split_uri.scheme,
+        uri_path=_unquote(split_uri.path),
+        user=_unquote(split_uri.username),
+        host=split_uri.hostname,
+        port=int(split_uri.port or DEFAULT_PORT),
+        password=_unquote(split_uri.password),
+    )
+
+
+def open_uri(uri, mode, transport_params):
+    smart_open.utils.check_kwargs(open, transport_params)
+    parsed_uri = parse_uri(uri)
+    uri_path = parsed_uri.pop('uri_path')
+    parsed_uri.pop('scheme')
+    return open(uri_path, mode, transport_params=transport_params, **parsed_uri)
+
+
+def convert_transport_params_to_args(transport_params):
+    supported_keywords = ['keyfile', 'certfile', 'context', 'timeout', 'source_address', 'encoding']
+    unsupported_keywords = [k for k in transport_params if k not in supported_keywords]
+    kwargs = {k: v for (k, v) in transport_params.items() if k in supported_keywords}
+
+    if unsupported_keywords:
+        logger.warning('ignoring unsupported ftp keyword arguments: %r', unsupported_keywords)
+
+    return kwargs
+
+
+def _connect(hostname, username, port, password, transport_params):
+    from ftplib import FTP_TLS
```

Comment:
Thank you for the PR. Have a look at how the other submodules are handling dependencies.

- https://github.com/RaRe-Technologies/smart_open/blob/4c6bc38d6b35aa75e5e9b8041bb80f0f956ac486/smart_open/azure.py#L19-L23
- https://github.com/RaRe-Technologies/smart_open/blob/develop/setup.py#L79

Reply:
Thank you for those resources. I have added the following code:
```
def _connect(hostname, username, port, password, transport_params):
    try:
        from ftplib import FTP_TLS
    except ImportError:
        raise Exception("Error: Unable to import FTP_TLS from ftplib library")
```

Let me know if there is anything else that I should add.

Reply:
No, that's not how we want to handle it. What we want to do io is:

1. Add an extra to setup.py to handle the new dependency (see example above)
2. In the new submodule, import the new dependency at module level, and if it fails, set MISSING_DEPS=True (see example above)

Reply:
Sorry about that, and thank you for bringing up my mistake. It turns out that ftplib is not a pip module and thus does not need to be installed separately (all python interpreters automatically have this library available for import). Thus, I have not added anything to setup.py, as technically there are no new dependencies introduced by my PR. However, I have changed `ftp.py` to import the ftplib library and appropriately set MISSING_DEPS at the module level as opposed to the function level, as you have brought up. Let me know if there are any other issues. And again, sorry for misunderstanding your earlier comment.

Reply:
Ah, I thought you're using some external library.

But if ftplib is part of the standard library, then how could it possibly be missing?

Reply:
That is a good point. I think that the missing dependencies test is unnecessary. I have removed it.